### ISSUE TITLE
How to configure istio On-Prem to complete Let's Encrypt Challenge?

### ISSUE THREAD

**Issue by bnussbau** [kubeflow/kubeflow#4866]
/kind question

**Question:**

 I modified the kctl_istio_dex.yaml (kubeflow version 1.0) file to include the letsencrypt overlay and set the parameters
```
- kustomizeConfig:
        overlays:
          - letsencrypt
          - self-signed
          - application
        parameters:
          - name: namespace
            value: cert-manager
          - name: acmeEmail
            value: $CERT_EMAIL
          - name: acmeUrl
            value: https://acme-v02.api.letsencrypt.org/directory
```
I see the letsencrypt cluster issuer, a certificate request (Message:  Waiting on certificate issuance from order kubeflow/ingress-cert-3687639054-858257646: "pending" ) , an Order with a challenge, an Ingress
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: istio
    nginx.ingress.kubernetes.io/whitelist-source-range: 0.0.0.0/0,::/0
  creationTimestamp: "2020-03-16T01:34:52Z"
  generateName: cm-acme-http-solver-
  generation: 1
  labels:
    acme.cert-manager.io/http-domain: "2669480700"
    acme.cert-manager.io/http-token: "1108807314"
    acme.cert-manager.io/http01-solver: "true"
  name: cm-acme-http-solver-bkhvs
  namespace: kubeflow
  ownerReferences:
  - apiVersion: cert-manager.io/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: Challenge
    name: ingress-cert-3687639054-858257646-3344660910
    uid: 53a71ef9-6726-11ea-b4fa-080027f4f707
  resourceVersion: "45428"
  selfLink: /apis/extensions/v1beta1/namespaces/kubeflow/ingresses/cm-acme-http-solver-bkhvs
  uid: 554c3327-6726-11ea-b4fa-080027f4f707
spec:
  rules:
  - host: $DOMAIN_NAME
    http:
      paths:
      - backend:
          serviceName: cm-acme-http-solver-7qgss
          servicePort: 8089
        path: /.well-known/acme-challenge/F0q4XSbFjom_qpYwATPMUB6nBXliq9jaK23XhknJ1Mk
status:
  loadBalancer: {}
```
My Gateway
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"networking.istio.io/v1alpha3","kind":"Gateway","metadata":{"annotations":{},"name":"kubeflow-gateway","namespace":"kubeflow"},"spec":{"selector":{"istio":"ingressgateway"},"servers":[{"hosts":["*"],"port":{"name":"http","number":80,"protocol":"HTTP"}}]}}
  creationTimestamp: "2020-03-15T23:55:13Z"
  generation: 13
  name: kubeflow-gateway
  namespace: kubeflow
  resourceVersion: "236464"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/kubeflow/gateways/kubeflow-gateway
  uid: 696a3405-6718-11ea-b4fa-080027f4f707
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: http
      number: 80
      protocol: HTTP
  - hosts:
    - $MY_DOMAIN
    port:
      name: https
      number: 443
      protocol: HTTPS
    tls:
      credentialName: ingress-cert
      mode: SIMPLE
      privateKey: sds
      serverCertificate: sds
```
Problem is that the Challenge cannot be completed as on port 80 requests are redirected to dex
` curl http://$domain/.well-known/acme-challenge/F0q4XSbFjom_qpYwATPMUB6nBXliq9jaK23XhknJ1Mk  `
`<a href="/dex/auth?client_id=kubeflow-oidc-authservice&amp;redirect_uri=%2Flogin%2Foidc&amp;response_type=code&amp;scope=profile+email+groups+openid&amp;state=MTU4NDM1NDI0OXxFd3dBRUd4RGQyTmhiakZxWlVORlFucFZjbkk9fMxHTeJSXTeeDkGbixqvk0tUr-gnKhLVc08eRsEVm5Ah">Found</a>.`
Thats also logged by the cert manager

```

11:34 Uhr
E0316 10:14:59.832196       1 sync.go:184] cert-manager/controller/challenges "msg"="propagation check failed" "error"="presented key (<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <title>dex</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link href="../static/main.css" rel="stylesheet">
    <link href="../theme/styles.css" rel="stylesheet">
    <link rel="icon" href="../theme/favicon.png">
  </head>
  <body class="theme-body">
    <div class="theme-navbar">
      <div class="theme-navbar__logo-wrap">
        <img class="theme-navbar__logo" src="../theme/logo.png">
      </div>
    </div>
    <div class="dex-container">
<div class="theme-panel">
  <h2 class="theme-heading">Log in to Your Account</h2>
  <form method="post" action="/dex/auth/local?req=gcobmqgsnorfne5ee5v34tka7">
    <div class="theme-form-row">
      <div class="theme-form-label">
        <label for="userid">Email Address</label>
      </div>
	  <input tabindex="1" required id="login" name="login" type="text" class="theme-form-input" placeholder="email address"  autofocus />
    </div>
    <div class="theme-form-row">
      <div class="theme-form-label">
        <label for="password">Password</label>
      </div>
	  <input tabindex="2" required id="password" name="password" type="password" class="theme-form-input" placeholder="password" />
    </div>
    <button tabindex="3" id="submit-login" type="submit" class="dex-btn theme-btn--primary">Login</button>
  </form>
</div>
    </div>
  </body>
</html>
) did not match expected (F0q4XSbFjom_qpYwATPMUBXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX)" "dnsName"="$DNS_NAME" "resource_kind"="Challenge" "resource_name"="ingress-cert-3687639054-858257646-3344660910" "resource_namespace"="kubeflow" "type"="http-01" 
I0316 10:14:59.978595       1 controller.go:135] cert-manager/controller/challen
```

How to expose the ingress so that the challenge can be completed? isio documentation (https://istio.io/docs/tasks/traffic-management/ingress/secure-ingress-sds/) also hasn’t helped me.
Can anyone help?

**Comment by jlewi** [comment]
If you are running on prem you most likely don't want to or can't use Lets Encrypt Challenge because that requires that your endpoint is publicly accessible which you most likely don't want to do.

You would most likely want to generate certificate some other way (e.g. a self-signed certificate) and then upload it to the cluster.

/cc @jtfogarty @johnugeorge @yanniszark

**Comment by bnussbau** [comment]
in that case the endpoint is publicly accessible, i'm just not able to configure istio correctly

**Comment by Oblynx** [comment]
I tried to go through the istio gateway but failed to route traffic to the `cm-acme-http-solver-` service. However this service is of type NodePort.
I ended up using my external load balancer to redirect traffic to <hostname>/.well-known to the node port of this service. It's definitely not an ideal solution, because I suspect auto-renewal won't work without reconfiguration.

**Comment by jbottum** [comment]
@bnussbau Hey Benjamin - Is this issue still open?   Did you have it resolved?  Can we close this?

**Comment by bnussbau** [comment]
@jbottum Hey, i was not able solve this yet. As a workaround I acquired the let's encrypt certificate via a local server outside from kubernetes and copied it manually. not an ideal solution. Thx @Oblynx for your workaroud description as well

**Comment by streamnsight** [comment]
This is an old issue but I struggled to find a solution so I'm posting here for those who will follow. 
I am using KF1.5.1
The main issue for letsencrypt to validate the http01 endpoint is that the `authservice` redirects everything to Dex.
So, following the crumbs trail, I found that authservice can whitelist paths to ignore.

in the `oidc-authservice/base/params.env`
set the 
```
SKIP_AUTH_URI=/dex /.well-known
```
Note, it is space separated and not comma separated as the doc says it should be. The doc is wrong.
Also, on newer versions of authservice, the ENV variable has been renamed to `SKIP_AUTH_URLS`, so it would be safe to add both in the authservice env to avoid issues in the future.
